src/proto/sim.rs

```diff
@@ -0,0 +1,1103 @@
+    pub fn run_while(&mut self, mut f: impl FnMut(PI, TopicId, Event<PI>) -> bool) {
+        'outer: loop {
+            while let Some((peer, topic, event)) = self.events.pop_front() {
+                if !f(peer, topic, event) {
+                    break 'outer;
+                }
+            }
+            self.tick();
+        }
+    }
+
+    /// Runs the simulation while `f` returns `true`, aborting after `timeout`.
+    ///
+    /// The callback will be called for each emitted event.
+    pub fn run_while_with_timeout(
+        &mut self,
+        timeout: Duration,
+        mut f: impl FnMut(PI, TopicId, Event<PI>) -> bool,
+    ) {
+        let end = self.time + timeout;
+        loop {
+            while let Some((peer, topic, event)) = self.events.pop_front() {
+                if !f(peer, topic, event) {
+                    return;
+                }
+            }
+            if self.queue.next_before(end) {
+                self.tick();
+            } else {
+                break;
+            }
+        }
+        assert!(self.time <= end);
+        self.time = end;
+    }
+
+    fn tick(&mut self) {
+        self.tick += 1;
+        let Some((time, peer, event)) = self.queue.pop() else {
+            warn!("tick on empty queue");
+            return;
+        };
+        assert!(time >= self.time);
+        self.time = time;
+        let span = debug_span!("tick", %peer, tick = %self.tick, t = %self.elapsed_fmt());
+        let _guard = span.enter();
+        debug!("~~ TICK ");
+
+        let Some(state) = self.peers.get_mut(&peer) else {
+            // TODO: queue PeerDisconnected for sender?
+            warn!(?time, ?peer, ?event, "event for dead peer");
+            return;
+        };
+        if let InEvent::RecvMessage(from, _message) = &event {
+            self.conns.insert((*from, peer).into());
+        }
+        let out = state.handle(event, self.time);
+        let mut kill = vec![];
+        for event in out {
+            match event {
+                OutEvent::SendMessage(to, message) => {
+                    let latency = latency_between(
+                        &self.config.latency,
+                        &mut self.latencies,
+                        &peer,
+                        &to,
+                        &mut self.rng,
+                    );
+                    self.queue
+                        .insert(self.time + latency, to, InEvent::RecvMessage(peer, message));
+                }
+                OutEvent::ScheduleTimer(time, timer) => {
+                    self.queue
+                        .insert(self.time + time, peer, InEvent::TimerExpired(timer));
+                }
+                OutEvent::DisconnectPeer(to) => {
+                    debug!(peer = ?peer, other = ?to, "disconnect");
+                    kill.push((peer, to));
+                }
+                OutEvent::EmitEvent(topic, event) => {
+                    debug!(peer = ?peer, "emit   {event:?}");
```

Comment:
```suggestion
                    debug!(peer = ?peer, "emit {event:?}");
```
unless it's intentional?

Reply:
Fixed